include/benchmark/benchmark.h

```diff
@@ -931,27 +919,30 @@ internal::Benchmark* RegisterBenchmark(const char* name, Lambda&& fn,
 #endif
 
 // The base class for all fixture tests.
-class Fixture : public internal::Benchmark {
+class Fixture {
  public:
-  Fixture() : internal::Benchmark("") {}
-
-  virtual void Run(State& st) {
-    this->SetUp(st);
-    this->BenchmarkCase(st);
-    this->TearDown(st);
-  }
-
+  virtual ~Fixture();
   // These will be deprecated ...
   virtual void SetUp(const State&) {}
   virtual void TearDown(const State&) {}
   // ... In favor of these.
   virtual void SetUp(State& st) { SetUp(const_cast<const State&>(st)); }
   virtual void TearDown(State& st) { TearDown(const_cast<const State&>(st)); }
 
- protected:
   virtual void BenchmarkCase(State&) = 0;
 };
 
+namespace internal {
+typedef Fixture* (FixtureCreator)();
+}
+
+// Create and register a fixture benchmark with the specified 'name' whose fixture
+// is created by creator
+//
+// RETURNS: A pointer to the registered benchmark.
+internal::Benchmark* RegisterBenchmark(const char* name,
```

Comment:
I'm not sure I agree in making this function part of the public API. Nobody should be calling this directly, right?

Reply:
This function is in the public api because there is this function https://github.com/google/benchmark/blob/master/include/benchmark/benchmark.h#L856 in the public api, I just keep them similar.  If users decide to write their own main function, and register benchmarks themselves, they may need this function. But class Benchmark itself is in the internal namespace, and I think it's wired for a function in the public api to return or accept parameters of types defined in the internal namespace.

Reply:
I will move the three RegisterBenchmark() functions(one for function, one for fixture, and if CXX11, one for lambda) to internal namespace.

Reply:
Please keep the lambda and function one as part of the public API.

The fixture overload works entirely differently that the existing ones, since you have to pass it a special function pointer *which isn't the benchmark, but a function which creates it*. The fact that this is how we handle registering fixtures feels like an internal implementation detail; not a part of the public API.

Reply:
> But class Benchmark itself is in the internal namespace,

Yeah, that's kind of dumb. I think the initial intent was that users were never supposed to name or create the type themselves; so the name was effectively "internal".